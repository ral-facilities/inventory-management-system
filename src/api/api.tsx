import axios from 'axios';
import { ErrorParsing, MicroFrontendId } from '../app.types';
import { readSciGatewayToken } from '../parseTokens';
import { settings } from '../settings';
import { InvalidateTokenType } from '../state/actions/actions.types';

export const imsApi = axios.create();

imsApi.interceptors.request.use(async (config) => {
  const apiUrl = (await settings)?.apiUrl || '';
  config.baseURL = apiUrl;
  config.headers['Authorization'] = `Bearer ${readSciGatewayToken()}`;
  return config;
});

// These are for ensuring refresh request is only sent once when multiple requests
// are failing due to 403's at the same time
let isFetchingAccessToken = false;
let failedAuthRequestQueue: ((shouldReject?: boolean) => void)[] = [];

/* This should be called when SciGateway successfully refreshes the access token - it retries
   all requests that failed due to an invalid token */
export const retryFailedAuthRequests = () => {
  isFetchingAccessToken = false;
  failedAuthRequestQueue.forEach((callback) => callback());
  failedAuthRequestQueue = [];
};

/* This should be called when SciGateway logs out as would occur if a token refresh fails
   due to the refresh token being out of date - it rejects all active request promises that
   were awaiting a token refresh using the orriginal error that occurred on the first attempt */
export const clearFailedAuthRequestsQueue = () => {
  isFetchingAccessToken = false;
  failedAuthRequestQueue.forEach((callback) => callback(true));
  failedAuthRequestQueue = [];
};

imsApi.interceptors.response.use(
  (response) => response,
  (error) => {
    const originalRequest = error.config;
    const errorMessage: string = error.response?.data
      ? (error.response.data as ErrorParsing).detail.toLocaleLowerCase() ??
        error.message
      : error.message;

    // Check if the token is invalid and needs refreshing
    // only allow a request to be retried once
    if (
      error.response?.status === 403 &&
      errorMessage.includes('expired token') &&
      !originalRequest._retried
    ) {
      originalRequest._retried = true;

      // Prevent other requests from also attempting to refresh while waiting for
      // SciGateway to refresh the token
      if (!isFetchingAccessToken) {
        isFetchingAccessToken = true;

        // Request SciGateway to refresh the token
        document.dispatchEvent(
          new CustomEvent(MicroFrontendId, {
            detail: {
              type: InvalidateTokenType,
            },
          })
        );
      }

      // Add request to queue to be resolved only once SciGateway has successfully
      // refreshed the token
      return new Promise((resolve, reject) => {
        failedAuthRequestQueue.push((shouldReject?: boolean) => {
          if (shouldReject) reject(error);
          else resolve(imsApi(originalRequest));
        });
      });
    }
    // Any other error
    else return Promise.reject(error);
  }
);
