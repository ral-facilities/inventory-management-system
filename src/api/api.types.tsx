interface CreatedModifiedMixin {
  created_time: string;
  modified_time: string;
}

export interface ErrorParsing {
  detail: string;
}

// Manufacturers

interface Address {
  address_line: string;
  town?: string | null;
  county?: string | null;
  country: string;
  postcode: string;
}

type AddressPatch = Partial<Address>;

export interface ManufacturerPost {
  name: string;
  url?: string | null;
  address: Address;
  telephone?: string | null;
}

export interface ManufacturerPatch
  extends Partial<Omit<ManufacturerPost, 'address'>> {
  address?: AddressPatch;
  id: string;
}

export interface Manufacturer extends ManufacturerPost, CreatedModifiedMixin {
  id: string;
  code: string;
}
