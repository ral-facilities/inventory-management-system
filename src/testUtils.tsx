import { LocalizationProvider } from '@mui/x-date-pickers';
import { AdapterDateFns } from '@mui/x-date-pickers/AdapterDateFnsV3';
import { QueryClient, QueryClientProvider } from '@tanstack/react-query';
import { RenderOptions, render } from '@testing-library/react';
import { enGB } from 'date-fns/locale/en-GB';
import React from 'react';
import {
  RouterProvider,
  createBrowserRouter,
  createMemoryRouter,
} from 'react-router-dom';
import { paths } from './App';
import {
  CatalogueCategory,
  CatalogueCategoryProperty,
  CatalogueItem,
  Item,
  Manufacturer,
} from './api/api.types';
import CatalogueCategoriesJSON from './mocks/CatalogueCategories.json';
import CatalogueItemsJSON from './mocks/CatalogueItems.json';
import ItemsJSON from './mocks/Items.json';
import ManufacturersJSON from './mocks/Manufacturers.json';

export const createTestQueryClient = (): QueryClient =>
  new QueryClient({
    defaultOptions: {
      queries: {
        retry: false,
        staleTime: 300000,
      },
    },
  });

interface ExtendedRenderOptions extends Omit<RenderOptions, 'queries'> {
  queryClient?: QueryClient;
}

function constructRouterProvider(
  ui: React.ReactNode,
  queryClient: QueryClient,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string
) {
  const Root: React.FunctionComponent = () => {
    return (
      <LocalizationProvider adapterLocale={enGB} dateAdapter={AdapterDateFns}>
        <QueryClientProvider client={queryClient}>{ui}</QueryClientProvider>
      </LocalizationProvider>
    );
  };

  const router =
    initialEntry !== undefined && urlPathKey !== undefined
      ? createMemoryRouter([{ path: paths[urlPathKey], Component: Root }], {
          initialEntries: [initialEntry],
        })
      : createBrowserRouter([{ path: '*', Component: Root }]);

  return { router: router, provider: <RouterProvider router={router} /> };
}

function constructRouterProviderWrapper(
  queryClient: QueryClient,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string
) {
  const wrapper = ({
    children,
  }: {
    children: React.ReactNode;
  }): JSX.Element => {
    return constructRouterProvider(
      children,
      queryClient,
      urlPathKey,
      initialEntry
    ).provider;
  };
  return wrapper;
}

export function renderComponentWithRouterProvider(
  ui: React.ReactElement,
  urlPathKey?: keyof typeof paths,
  initialEntry?: string,
  {
    // Automatically create a query client instance if no query client was passed in
    queryClient = createTestQueryClient(),
    ...renderOptions
  }: ExtendedRenderOptions = {}
) {
  const { router, provider } = constructRouterProvider(
    ui,
    queryClient,
    urlPathKey,
    initialEntry
  );
  return {
    queryClient,
    router,
    ...render(ui, {
      wrapper: () => provider,
      ...renderOptions,
    }),
  };
}

export const hooksWrapperWithProviders = (props?: {
  queryClient?: QueryClient;
  urlPathKey?: keyof typeof paths;
  initialEntry?: string;
}) => {
  const testQueryClient = props?.queryClient ?? createTestQueryClient();
  return constructRouterProviderWrapper(
    testQueryClient,
    props?.urlPathKey,
    props?.initialEntry
  );
};

export const getCatalogueItemsPropertiesById = (
  id: string
): CatalogueCategoryProperty[] => {
  const filteredCategories = CatalogueCategoriesJSON.filter(
    (catalogueCategory) => catalogueCategory.id === id
  );

  if (filteredCategories.length === 0) {
    return [];
  }

  const properties = filteredCategories[0].properties ?? [];
  return properties as CatalogueCategoryProperty[];
};

export const getCatalogueCategoryById = (
  id: string
): CatalogueCategory | undefined => {
  return (
    (CatalogueCategoriesJSON.find(
      (catalogueCategory) => catalogueCategory.id === id
    ) as CatalogueCategory) || undefined
  );
};

export const getCatalogueItemById = (id: string): CatalogueItem | undefined => {
  return (
    (CatalogueItemsJSON.find(
      (catalogueItem) => catalogueItem.id === id
    ) as CatalogueItem) || undefined
  );
};

export const catalogueItemData = (id: string): CatalogueItem[] => {
  return CatalogueItemsJSON.filter(
    (catalogueItem) => catalogueItem.catalogue_category_id === id
  );
};

export const getManufacturerById = (id: string): Manufacturer | undefined => {
  return (
    (ManufacturersJSON.find(
      (manufacturer) => manufacturer.id === id
    ) as Manufacturer) || undefined
  );
};

export const getItemsByCatalogueItemId = (catalogueItemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.catalogue_item_id === catalogueItemId);
};

export const getItemById = (itemId: string): Item | undefined => {
  return ItemsJSON.find((item) => item.id === itemId);
};

export const getItemsBySystemId = (systemId: string): Item[] => {
  return ItemsJSON.filter((item) => item.system_id === systemId);
};

export const CREATED_MODIFIED_TIME_VALUES = {
  created_time: '2024-01-01T12:00:00.000+00:00',
  modified_time: '2024-01-02T13:10:10.000+00:00',
};
