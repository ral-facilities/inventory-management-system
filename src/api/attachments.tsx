import { useMutation, UseMutationResult } from '@tanstack/react-query';
import { AxiosError } from 'axios';
import { storageApi } from './api';
import {
  AttachmentsPostMetadata,
  AttachmentsPostMetadataResponse,
} from './api.types';

const postAttachmentMetadata = async (
  attachmentMetadata: AttachmentsPostMetadata
): Promise<AttachmentsPostMetadataResponse> => {
  return storageApi
    .post<AttachmentsPostMetadataResponse>(`/attachments`, attachmentMetadata)
    .then((response) => response.data);
};

export const usePostAttachmentMetadata = (): UseMutationResult<
  AttachmentsPostMetadataResponse,
  AxiosError,
  AttachmentsPostMetadata
> => {
  return useMutation({
    mutationFn: (attachmentMetadata: AttachmentsPostMetadata) =>
      postAttachmentMetadata(attachmentMetadata),
  });
};
