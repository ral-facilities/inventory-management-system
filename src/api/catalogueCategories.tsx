import {
  useMutation,
  UseMutationResult,
  useQuery,
  useQueryClient,
  UseQueryResult,
} from '@tanstack/react-query';
import { AxiosError } from 'axios';
import {
  AddCatalogueCategory,
  AddPropertyMigration,
  BreadcrumbsInfo,
  CatalogueCategory,
  CatalogueCategoryProperty,
  CopyToCatalogueCategory,
  EditCatalogueCategory,
  ErrorParsing,
  MoveToCatalogueCategory,
  EditPropertyMigration,
  TransferState,
} from '../app.types';

import handleTransferState from '../handleTransferState';
import { generateUniqueName } from '../utils';
import { imsApi } from './api';

const fetchCatalogueCategories = async (
  parent_id: string
): Promise<CatalogueCategory[]> => {
  const queryParams = new URLSearchParams();

  queryParams.append('parent_id', parent_id);

  return imsApi
    .get(`/v1/catalogue-categories`, {
      params: queryParams,
    })
    .then((response) => {
      return response.data;
    });
};

export const useCatalogueCategories = (
  isLeaf: boolean,
  parent_id: string
): UseQueryResult<CatalogueCategory[], AxiosError> => {
  return useQuery({
    queryKey: ['CatalogueCategories', parent_id],
    queryFn: () => {
      return fetchCatalogueCategories(parent_id);
    },
    enabled: !isLeaf,
  });
};

const fetchCatalogueBreadcrumbs = async (
  id: string
): Promise<BreadcrumbsInfo> => {
  return imsApi
    .get(`/v1/catalogue-categories/${id}/breadcrumbs`, {})
    .then((response) => {
      return response.data;
    });
};

export const useCatalogueBreadcrumbs = (
  id?: string | null
): UseQueryResult<BreadcrumbsInfo, AxiosError> => {
  return useQuery({
    queryKey: ['CatalogueBreadcrumbs', id],
    queryFn: () => {
      return fetchCatalogueBreadcrumbs(id ?? '');
    },
    enabled: !!id,
  });
};

const addCatalogueCategory = async (
  catalogueCategory: AddCatalogueCategory
): Promise<CatalogueCategory> => {
  return imsApi
    .post<CatalogueCategory>(`/v1/catalogue-categories`, catalogueCategory)
    .then((response) => response.data);
};

export const useAddCatalogueCategory = (): UseMutationResult<
  CatalogueCategory,
  AxiosError,
  AddCatalogueCategory
> => {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (catalogueCategory: AddCatalogueCategory) =>
      addCatalogueCategory(catalogueCategory),
    onSuccess: (category) => {
      queryClient.invalidateQueries({
        queryKey: ['CatalogueCategories', category.parent_id ?? 'null'],
      });
    },
  });
};

const addCatalogueCategoryProperty = async (
  addPropertyMigration: AddPropertyMigration
): Promise<CatalogueCategoryProperty> => {
  return imsApi
    .post<CatalogueCategoryProperty>(
      `/v1/catalogue-categories/${addPropertyMigration.catalogueCategory.id}/properties`,
      addPropertyMigration.property
    )
    .then((response) => response.data);
};

export const useAddCatalogueCategoryProperty = (): UseMutationResult<
  CatalogueCategoryProperty,
  AxiosError,
  AddPropertyMigration
> => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (addPropertyMigration: AddPropertyMigration) => {
      handleTransferState([
        {
          name: addPropertyMigration.catalogueCategory.name,
          message: `Adding property ${addPropertyMigration.property.name} in ${addPropertyMigration.catalogueCategory.name}`,
          state: 'information',
        },
      ]);
      return addCatalogueCategoryProperty(addPropertyMigration);
    },
    onSuccess: (data, variables) => {
      const { name } = data;
      queryClient.invalidateQueries({
        queryKey: [
          'CatalogueCategories',
          variables.catalogueCategory.parent_id,
        ],
      });
      queryClient.invalidateQueries({
        queryKey: ['CatalogueItems', variables.catalogueCategory.id],
      });
      queryClient.invalidateQueries({
        queryKey: ['Items'],
      });

      handleTransferState([
        {
          name: variables.catalogueCategory.name,
          message: `Successfully added property ${name} in ${variables.catalogueCategory.name}`,
          state: 'success',
        },
      ]);
    },
    onError: (error, variables) => {
      const response = error.response?.data as ErrorParsing;

      handleTransferState([
        {
          name: variables.catalogueCategory.name,
          message: response.detail,
          state: 'error',
        },
      ]);
    },
  });
};

const editCatalogueCategoryProperty = async (
  editPropertyMigration: EditPropertyMigration
): Promise<CatalogueCategoryProperty> => {
  const { id, ...propertyBody } = editPropertyMigration.property;
  return imsApi
    .patch<CatalogueCategoryProperty>(
      `/v1/catalogue-categories/${editPropertyMigration.catalogueCategory.id}/properties/${editPropertyMigration.property.id}`,
      propertyBody
    )
    .then((response) => response.data);
};

export const useEditCatalogueCategoryProperty = (): UseMutationResult<
  CatalogueCategoryProperty,
  AxiosError,
  EditPropertyMigration
> => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: (editPropertyMigration: EditPropertyMigration) => {
      handleTransferState([
        {
          name: editPropertyMigration.catalogueCategory.name,
          message: `Editing property ${editPropertyMigration.property.name} in ${editPropertyMigration.catalogueCategory.name}`,
          state: 'information',
        },
      ]);
      return editCatalogueCategoryProperty(editPropertyMigration);
    },
    onSuccess: (data, variables) => {
      const { name } = data;
      queryClient.invalidateQueries({
        queryKey: [
          'CatalogueCategories',
          variables.catalogueCategory.parent_id,
        ],
      });
      queryClient.invalidateQueries({
        queryKey: ['CatalogueItems', variables.catalogueCategory.id],
      });
      queryClient.invalidateQueries({
        queryKey: ['Items'],
      });
      handleTransferState([
        {
          name: variables.catalogueCategory.name,
          message: `Successfully edited property ${name} in ${variables.catalogueCategory.name}`,
          state: 'success',
        },
      ]);
    },
    onError: (error, variables) => {
      const response = error.response?.data as ErrorParsing;

      handleTransferState([
        {
          name: variables.catalogueCategory.name,
          message: response.detail,
          state: 'error',
        },
      ]);
    },
  });
};

const editCatalogueCategory = async (
  catalogueCategory: EditCatalogueCategory
): Promise<CatalogueCategory> => {
  const { id, ...updatedCategory } = catalogueCategory;
  return imsApi
    .patch<CatalogueCategory>(`/v1/catalogue-categories/${id}`, updatedCategory)
    .then((response) => response.data);
};

export const useEditCatalogueCategory = (): UseMutationResult<
  CatalogueCategory,
  AxiosError,
  EditCatalogueCategory
> => {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (catalogueCategory: EditCatalogueCategory) =>
      editCatalogueCategory(catalogueCategory),
    onSuccess: (category) => {
      queryClient.invalidateQueries({
        queryKey: ['CatalogueCategories', category.parent_id ?? 'null'],
      });
      queryClient.invalidateQueries({
        // Don't use ID here as will also need to update any of its children as well
        queryKey: ['CatalogueBreadcrumbs'],
      });
      queryClient.invalidateQueries({
        queryKey: ['CatalogueCategory', category.id],
      });
    },
  });
};

export const useMoveToCatalogueCategory = (): UseMutationResult<
  TransferState[],
  AxiosError,
  MoveToCatalogueCategory
> => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (moveToCatalogueCategory: MoveToCatalogueCategory) => {
      const transferStates: TransferState[] = [];

      // Ids for invalidation (parentIds must be a string value of 'null' for invalidation)
      const successfulIds: string[] = [];
      const successfulParentIds: string[] = [];

      const promises = moveToCatalogueCategory.selectedCategories.map(
        async (category: CatalogueCategory) => {
          return editCatalogueCategory({
            id: category.id,
            parent_id: moveToCatalogueCategory.targetCategory?.id || null,
          })
            .then((result) => {
              transferStates.push({
                name: result.name ?? '',
                message: `Successfully moved to ${
                  moveToCatalogueCategory.targetCategory?.name || 'Root'
                }`,
                state: 'success',
              });

              successfulIds.push(category.id);
              successfulParentIds.push(category.parent_id || 'null');
            })
            .catch((error) => {
              const response = error.response?.data as ErrorParsing;

              transferStates.push({
                name: category.name,
                message: response.detail,
                state: 'error',
              });
            });
        }
      );

      await Promise.all(promises);

      if (successfulIds.length > 0) {
        queryClient.invalidateQueries({
          queryKey: [
            'CatalogueCategories',
            moveToCatalogueCategory.targetCategory?.id || 'null',
          ],
        });
        // Also need to invalidate each parent we are moving from (likely just the one)
        const uniqueParentIds = new Set(successfulParentIds);
        uniqueParentIds.forEach((parentId: string) =>
          queryClient.invalidateQueries({
            queryKey: ['CatalogueCategories', parentId],
          })
        );
        queryClient.invalidateQueries({ queryKey: ['CatalogueBreadcrumbs'] });
        successfulIds.forEach((id: string) =>
          queryClient.invalidateQueries({
            queryKey: ['CatalogueCategory', id],
          })
        );
      }

      return transferStates;
    },
  });
};

export const useCopyToCatalogueCategory = (): UseMutationResult<
  TransferState[],
  AxiosError,
  CopyToCatalogueCategory
> => {
  const queryClient = useQueryClient();

  return useMutation({
    mutationFn: async (copyToCatalogueCategory: CopyToCatalogueCategory) => {
      const transferStates: TransferState[] = [];

      // Ids for invalidation (must be a string value of 'null' if null
      // for invalidation)
      const successfulParentIds: string[] = [];

      const promises = copyToCatalogueCategory.selectedCategories.map(
        async (category: CatalogueCategory) => {
          // Data to post (backend will just ignore the extra here - only id and code)
          // Also use Object.assign to copy the data otherwise will modify in place causing issues
          // in tests
          const categoryAdd: AddCatalogueCategory = Object.assign(
            {},
            category
          ) as AddCatalogueCategory;

          // Assign new parent
          categoryAdd.parent_id =
            copyToCatalogueCategory.targetCategory?.id || null;

          // Avoid duplicates
          categoryAdd.name = generateUniqueName(
            categoryAdd.name,
            copyToCatalogueCategory.existingCategoryNames
          );

          return addCatalogueCategory(categoryAdd)
            .then((result) => {
              const targetCategoryName =
                copyToCatalogueCategory.targetCategory?.name || 'Root';
              transferStates.push({
                name: result.name,
                message: `Successfully copied to ${targetCategoryName}`,
                state: 'success',
              });

              successfulParentIds.push(result.parent_id || 'null');
            })
            .catch((error) => {
              const response = error.response?.data as ErrorParsing;

              transferStates.push({
                name: category.name,
                message: response.detail,
                state: 'error',
              });
            });
        }
      );

      await Promise.all(promises);

      if (successfulParentIds.length > 0)
        // Only invalidate the unique parents
        new Set(successfulParentIds).forEach((parentId) =>
          queryClient.invalidateQueries({
            queryKey: ['CatalogueCategories', parentId],
          })
        );

      return transferStates;
    },
  });
};
const deleteCatalogueCategory = async (
  catalogueCategory: CatalogueCategory
): Promise<void> => {
  return imsApi
    .delete(`/v1/catalogue-categories/${catalogueCategory.id}`)
    .then((response) => response.data);
};

export const useDeleteCatalogueCategory = (): UseMutationResult<
  void,
  AxiosError,
  CatalogueCategory
> => {
  const queryClient = useQueryClient();
  return useMutation({
    mutationFn: (catalogueCategory: CatalogueCategory) =>
      deleteCatalogueCategory(catalogueCategory),
    onSuccess: () => {
      queryClient.invalidateQueries({ queryKey: ['CatalogueCategories'] });
    },
  });
};

const fetchCatalogueCategory = async (
  id: string | undefined
): Promise<CatalogueCategory> => {
  return imsApi.get(`/v1/catalogue-categories/${id}`, {}).then((response) => {
    return response.data;
  });
};

export const useCatalogueCategory = (
  id?: string | null
): UseQueryResult<CatalogueCategory, AxiosError> => {
  return useQuery({
    queryKey: ['CatalogueCategory', id],
    queryFn: () => {
      return fetchCatalogueCategory(id ?? '');
    },
    enabled: !!id,
  });
};
