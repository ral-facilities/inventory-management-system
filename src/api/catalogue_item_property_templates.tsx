import { UseQueryResult, useQuery } from '@tanstack/react-query';
import { AxiosError } from 'axios';
import { CatalogueItemPropertyTemplate } from '../app.types';
import { imsApi } from './api';

const fetchCatalogueItemPropertyTemplates = async (): Promise<
  CatalogueItemPropertyTemplate[]
> => {
  return imsApi
    .get('/v1/catalogue-item-property-templates')
    .then((response) => {
      return response.data;
    });
};

export const useCatalogueItemPropertyTemplates = (): UseQueryResult<
  CatalogueItemPropertyTemplate[],
  AxiosError
> => {
  return useQuery({
    queryKey: ['CatalogueItemPropertyTemplates'],
    queryFn: () => {
      return fetchCatalogueItemPropertyTemplates();
    },
  });
};
