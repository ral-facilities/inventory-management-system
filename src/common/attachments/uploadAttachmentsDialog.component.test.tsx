import { fireEvent, screen, waitFor } from '@testing-library/react';
import userEvent, { UserEvent } from '@testing-library/user-event';
import { act } from 'react';
import { renderComponentWithRouterProvider } from '../../testUtils';

import { MockInstance } from 'vitest';
import { storageApi } from '../../api/api';
import UploadAttachmentsDialog, {
  UploadAttachmentsDialogProps,
} from './uploadAttachmentsDialog.component';

//
describe('Upload attachment dialog', () => {
  let props: UploadAttachmentsDialogProps;
  let user: UserEvent;
  let axiosPostSpy: MockInstance;
  let xhrPostSpy: MockInstance;

  const onClose = vi.fn();

  const createView = () => {
    return renderComponentWithRouterProvider(
      <UploadAttachmentsDialog {...props} />
    );
  };

  beforeEach(() => {
    props = {
      open: true,
      onClose: onClose,
      entityId: '1',
    };
    user = userEvent.setup();
    axiosPostSpy = vi.spyOn(storageApi, 'post');
    xhrPostSpy = vi.spyOn(window.XMLHttpRequest.prototype, 'open');
    vi.spyOn(console, 'error').mockImplementation(() => {});
  });

  afterEach(() => {
    vi.clearAllMocks();
    axiosPostSpy.mockRestore();
    xhrPostSpy.mockRestore();
  });

  it('renders dialog correctly', async () => {
    let baseElement;
    await act(async () => {
      baseElement = createView().baseElement;
    });

    expect(
      screen.getByText('Files cannot be larger than 100MB')
    ).toBeInTheDocument();
    expect(baseElement).toMatchSnapshot();
  });

  it('calls onclose when close button is clicked', async () => {
    createView();

    await user.click(screen.getByRole('button', { name: 'Close' }));

    expect(onClose).toHaveBeenCalled();
  });

  it('posts attachment metadata successfully', async () => {
    // Render the component

    createView();

    const file1 = new File(['test'], 'test1.txt', {
      type: 'text/plain',
    });

    // Find the Uppy Dashboard's drop zone (it usually has a label like "Drop files here" or "Browse files")
    const dropZone = screen.getByText(/files cannot be larger than/i);

    // Create a drag-and-drop event for the file
    Object.defineProperty(dropZone, 'files', {
      value: [file1],
    });

    // Fire the drop event

    fireEvent.drop(dropZone, {
      dataTransfer: {
        files: [file1],
      },
    });

    // Wait for the UI to update with the added file
    await waitFor(() => {
      expect(screen.getByText('test1.txt')).toBeInTheDocument();
    });

    await user.click(await screen.findByText('Upload 1 file'));

    // Assert axios post was called
    expect(axiosPostSpy).toHaveBeenCalledWith('/attachments', {
      description: '',
      entity_id: '1',
      file_name: 'test1.txt',
      title: '',
    });

    expect(xhrPostSpy).toHaveBeenCalledWith(
      'POST',
      'object-storage/test1',
      true
    );
  });
  // Works locally but doesn't work on CI
  it.skip('errors when presigned url fails', async () => {
    // Render the component

    createView();

    const file1 = new File(['test'], 'uploadError.txt', {
      type: 'text/plain',
    });

    // Find the Uppy Dashboard's drop zone (it usually has a label like "Drop files here" or "Browse files")
    const dropZone = screen.getByText(/files cannot be larger than/i);

    // Create a drag-and-drop event for the file
    Object.defineProperty(dropZone, 'files', {
      value: [file1],
    });

    // Fire the drop event

    fireEvent.drop(dropZone, {
      dataTransfer: {
        files: [file1],
      },
    });

    // Wait for the UI to update with the added file
    await waitFor(() => {
      expect(screen.getByText('uploadError.txt')).toBeInTheDocument();
    });

    await user.click(await screen.findByText('Upload 1 file'));

    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'Close' })).toBeDisabled();
    });

    // Assert axios post was called
    expect(axiosPostSpy).toHaveBeenCalledWith('/attachments', {
      description: '',
      entity_id: '1',
      file_name: 'uploadError.txt',
      title: '',
    });

    expect(xhrPostSpy).toHaveBeenCalledWith(
      'POST',
      'object-storage/uploadError',
      true
    );
    await waitFor(() => {
      expect(screen.getByRole('button', { name: 'Close' })).not.toBeDisabled();
    });
    expect(await screen.findByText('Upload failed')).toBeInTheDocument();
    expect(
      await screen.findByLabelText('Show error details')
    ).toBeInTheDocument();
  }, 20000);

  it('errors when file is removed mid upload', async () => {
    // Render the component

    createView();

    const file1 = new File(['test'], 'removeError.txt', {
      type: 'text/plain',
    });

    // Find the Uppy Dashboard's drop zone (it usually has a label like "Drop files here" or "Browse files")
    const dropZone = screen.getByText(/files cannot be larger than/i);

    // Create a drag-and-drop event for the file
    Object.defineProperty(dropZone, 'files', {
      value: [file1],
    });

    // Fire the drop event

    fireEvent.drop(dropZone, {
      dataTransfer: {
        files: [file1],
      },
    });

    // Wait for the UI to update with the added file
    await waitFor(() => {
      expect(screen.getByText('removeError.txt')).toBeInTheDocument();
    });

    await user.click(await screen.findByText('Upload 1 file'));

    // Assert axios post was called
    expect(axiosPostSpy).toHaveBeenCalledWith('/attachments', {
      description: '',
      entity_id: '1',
      file_name: 'removeError.txt',
      title: '',
    });

    expect(xhrPostSpy).toHaveBeenCalledWith(
      'POST',
      'object-storage/removeError',
      true
    );

    await user.click(
      await screen.findByRole('button', { name: 'Remove file' })
    );

    expect(screen.queryByText('Upload 1 file')).not.toBeInTheDocument();
  });
});
