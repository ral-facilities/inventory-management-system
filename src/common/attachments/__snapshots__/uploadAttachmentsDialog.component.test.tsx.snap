// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Upload attachment dialog > renders dialog correctly 1`] = `
<body
  style="padding-right: 1024px; overflow: hidden;"
>
  <div
    aria-hidden="true"
  />
  <div
    class="MuiDialog-root MuiModal-root css-zw3mfo-MuiModal-root-MuiDialog-root"
    role="presentation"
  >
    <div
      aria-hidden="true"
      class="MuiBackdrop-root MuiModal-backdrop css-yiavyu-MuiBackdrop-root-MuiDialog-backdrop"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
    />
    <div
      data-testid="sentinelStart"
      tabindex="0"
    />
    <div
      class="MuiDialog-container MuiDialog-scrollPaper css-hz1bth-MuiDialog-container"
      role="presentation"
      style="opacity: 1; webkit-transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms; transition: opacity 225ms cubic-bezier(0.4, 0, 0.2, 1) 0ms;"
      tabindex="-1"
    >
      <div
        aria-labelledby=":r0:"
        class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation24 MuiDialog-paper MuiDialog-paperScrollPaper MuiDialog-paperWidthMd MuiDialog-paperFullWidth css-rnmm7m-MuiPaper-root-MuiDialog-paper"
        role="dialog"
      >
        <h2
          class="MuiTypography-root MuiTypography-h6 MuiDialogTitle-root css-bdhsul-MuiTypography-root-MuiDialogTitle-root"
          id=":r0:"
        >
          Upload Attachments
        </h2>
        <div
          class="MuiDialogContent-root css-thk4mm-MuiDialogContent-root"
        >
          <div
            class="uppy-Container"
          >
            <div
              class="uppy-Root"
              dir="ltr"
            >
              <div
                aria-disabled="false"
                aria-hidden="false"
                aria-label="Uppy Dashboard"
                class="uppy-Dashboard uppy-Dashboard--animateOpenClose"
                data-uppy-drag-drop-supported="true"
                data-uppy-num-acquirers="0"
                data-uppy-theme="light"
              >
                <div
                  aria-hidden="true"
                  class="uppy-Dashboard-overlay"
                  tabindex="-1"
                />
                <div
                  aria-modal="false"
                  class="uppy-Dashboard-inner"
                  style="width: 750px; height: 550px;"
                >
                  <div
                    class="uppy-Dashboard-innerWrap"
                  >
                    <div
                      class="uppy-Dashboard-dropFilesHereHint"
                    >
                      Drop your files here
                    </div>
                    <div
                      class="uppy-Dashboard-AddFiles"
                    >
                      <input
                        aria-hidden="true"
                        class="uppy-Dashboard-input"
                        hidden=""
                        multiple=""
                        name="files[]"
                        tabindex="-1"
                        type="file"
                      />
                      <input
                        aria-hidden="true"
                        class="uppy-Dashboard-input"
                        hidden=""
                        multiple=""
                        name="files[]"
                        tabindex="-1"
                        type="file"
                        webkitdirectory="true"
                      />
                      <div
                        class="uppy-Dashboard-AddFiles-title"
                      >
                        Drop files here or 
                        <button
                          class="uppy-u-reset uppy-c-btn uppy-Dashboard-browse"
                          data-uppy-super-focusable="true"
                          type="button"
                        >
                          browse files
                        </button>
                      </div>
                      <div
                        class="uppy-Dashboard-AddFiles-list"
                        role="tablist"
                      >
                        <span
                          role="presentation"
                          style="white-space: nowrap;"
                        />
                      </div>
                      <div
                        class="uppy-Dashboard-AddFiles-info"
                      >
                        <div
                          class="uppy-Dashboard-note"
                        >
                          Files cannot be larger than 100MB
                        </div>
                      </div>
                    </div>
                    <div
                      class="uppy-Dashboard-progressindicators"
                    >
                      <div
                        aria-hidden="true"
                        class="uppy-StatusBar is-waiting"
                      >
                        <div
                          aria-label="0%"
                          aria-valuemax="100"
                          aria-valuemin="0"
                          aria-valuenow="0"
                          aria-valuetext="0%"
                          class="uppy-StatusBar-progress"
                          role="progressbar"
                          style="width: 0%;"
                        />
                        <div
                          class="uppy-StatusBar-actions"
                        />
                      </div>
                      <div
                        class="uppy uppy-Informer"
                      >
                        <span />
                      </div>
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
        <div
          class="MuiDialogActions-root MuiDialogActions-spacing css-knqc4i-MuiDialogActions-root"
        >
          <button
            class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorSecondary MuiButton-root MuiButton-text MuiButton-textSecondary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorSecondary css-12k9yya-MuiButtonBase-root-MuiButton-root"
            tabindex="0"
            type="button"
          >
            Close
            <span
              class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
            />
          </button>
        </div>
      </div>
    </div>
    <div
      data-testid="sentinelEnd"
      tabindex="0"
    />
  </div>
</body>
`;
